src/NuGet.Core/NuGet.Protocol.Core.v3/Resources/ServiceIndexResourceV3.cs

```diff
@@ -14,129 +16,210 @@ namespace NuGet.Protocol
+        private readonly IDictionary<string, List<ServiceIndexEntry>> _index;
         private readonly DateTime _requestTime;
-        private List<Uri> _empty;
+        private static readonly IReadOnlyList<ServiceIndexEntry> _emptyEntries = new List<ServiceIndexEntry>();
+        private static readonly IReadOnlyList<Uri> _emptyUris = new List<Uri>();
+        private static readonly SemanticVersion _defaultVersion = new SemanticVersion(0, 0, 0);
 
         public ServiceIndexResourceV3(JObject index, DateTime requestTime)
         {
             _index = MakeLookup(index);
             _requestTime = requestTime;
         }
 
-        private static IDictionary<string, List<Uri>> MakeLookup(JObject index)
+        /// <summary>
+        /// Time the index was requested
+        /// </summary>
+        public virtual DateTime RequestTime
         {
-            var result = new Dictionary<string, List<Uri>>();
+            get { return _requestTime; }
+        }
 
-            JToken resources;
-            if (index.TryGetValue("resources", out resources))
+        /// <summary>
+        /// All service index entries.
+        /// </summary>
+        public virtual IReadOnlyList<ServiceIndexEntry> Entries
+        {
+            get
             {
-                foreach (var resource in resources)
-                {
-                    JToken type = resource["@type"];
-                    JToken id = resource["@id"];
+                return _index.SelectMany(e => e.Value).ToList();
+            }
+        }
 
-                    if (type == null || id == null)
-                    {
-                        continue;
-                    }
+        /// <summary>
+        /// Get the list of service entries that best match the current clientVersion and type.
+        /// </summary>
+        public virtual IReadOnlyList<ServiceIndexEntry> GetServiceEntries(params string[] orderedTypes)
+        {
+            var clientVersion = MinClientVersionUtility.GetNuGetClientVersion();
 
-                    if (type.Type == JTokenType.Array)
-                    {
-                        foreach (var nType in type)
-                        {
-                            AddEndpoint(result, nType, id);
-                        }
-                    }
-                    else
+            return GetServiceEntries(clientVersion, orderedTypes);
+        }
+
+        /// <summary>
+        /// Get the list of service entries that best match the clientVersion and type.
+        /// </summary>
+        public virtual IReadOnlyList<ServiceIndexEntry> GetServiceEntries(NuGetVersion clientVersion, params string[] orderedTypes)
+        {
+            if (clientVersion == null)
+            {
+                throw new ArgumentNullException(nameof(clientVersion));
+            }
+
+            foreach (var type in orderedTypes)
+            {
+                List<ServiceIndexEntry> entries;
+                if (_index.TryGetValue(type, out entries))
```

Comment:
not: `List<ServiceIndexEntry> entries` do not need a separate declaration. they can be part of the function call. I can give you an example if needed.